docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,491 @@
+  BlogPost,
+  (DataSource, props) => DataSource.getBlogPost(props.id)
+});
+```
+
+The first parameter is the wrapped component. The second parameter retrieves the data we're interested in, given a `DataSource` and the current props.
+
+When `CommentListWithSubscription` and `BlogPostWithSubscription` are rendered, `CommentList` and `BlogPost` will be passed a `data` prop with the most current data retrieved from `DataSource`:
+
+```js
+// This function takes a component...
+function withSubscription(WrappedComponent, selectData) {
+  // ...and returns another component...
+  return class extends React.Component {
+    constructor(props) {
+      super();
+      this.handleChange = this.handleChange.bind(this);
+      this.state = {
+        data: selectData(DataSource, props)
+      };
+    }
+
+    componentDidMount() {
+      // ... that takes care of the subscription...
+      DataSource.addChangeListener(this.handleChange);
+    }
+
+    componentWillUnmount() {
+      DataSource.removeChangeListener(this.handleChange);
+    }
+
+    handleChange() {
+      this.setState({
+        comments: selectData(DataSource, this.props)
+      });
+    }
+
+    render: function() {
+      // ... and renders the wrapped component with the fresh data!
+      // Notice that we pass through any additional props
+      return <WrappedComponent data={this.state.data} {...this.props} />;
+    }
+  });
+}
+```
+
+Note that an HOC doesn't modify the input component, nor does it use inheritance to copy its behavior. Rather, an HOC *composes* the original component by *wrapping* it in a container component. An HOC is a pure function with zero side-effects.
+
+And that's it! The wrapped component receives all the props of the container, along with a new prop, `data`, which it uses to render its output. The HOC isn't concerned with how or why the data is used, and the wrapped component isn't concerned with where the data came from.
+
+Because `withSubscription` is a normal function, you can add as many or as few arguments as you like. For example, you may want to make the name of the `data` prop configurable, to further isolate the HOC from the wrapped component. Or you could accept an argument that configures `shouldComponentUpdate`, or one that configures the data source. These are all possible because the HOC has full control over how the component is defined.
+
+Like components, the contract between `withSubscription` and the wrapped component is entirely props-based. This makes it easy to swap one HOC for a different one, as long as they provide the same props to the wrapped component. This may be useful if you change data-fetching libraries, for example.
+
+### Why Not Use a Component?
+
+Components are a good choice when you want to vary the output based on the given props. However, in our example, the difference between `CommentList` and `BlogPost` isn't limited to their output. They also differ in the implementation of their lifecycle methods.
```

Comment:
> The reason to not use a component seems simple to me. A component can't mess with its childrens' state. A HOC can totally mess with the state of the wrapped component.

That doesn't seem clarifying to me at all, to be honest.

I think you're right that I can probably just scrap this section. It was a bit awkward to write anyway, because there is a vocal contingent of folks who prefer components + render callbacks to HOCs. But perhaps that belongs in a separate doc that compares the two approaches.